app/uk/gov/hmrc/transitmovementsrouter/controllers/MessagesController.scala

```diff
@@ -177,30 +173,64 @@ class MessagesController @Inject() (
       )
     ).toMovementAndMessageId
 
-  private def updateStatus(movementId: MovementId, messageId: MessageId, messageStatus: MessageStatus)(implicit
+  private def updateAndSendNotification(movementId: MovementId, messageId: MessageId, messageStatus: MessageStatus, jsonValue: JsValue)(implicit
     hc: HeaderCarrier
   ): EitherT[Future, PresentationError, Unit] =
-    persistenceConnector
-      .patchMessageStatus(movementId, messageId, MessageUpdate(messageStatus))
-      .asPresentation
+    for {
+      persistenceResponse <- persistenceConnector
+        .patchMessageStatus(movementId, messageId, MessageUpdate(messageStatus))
+        .asPresentation
+      _ = pushNotificationsConnector
+        .postJSON(
+          movementId,
+          messageId,
+          jsonValue
+        )
+    } yield persistenceResponse
 
   def handleSdesResponse() =
     Action.async(parse.json) {
       implicit request =>
         implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromRequest(request)
-        (for {
-          sdesResponse <- parseAndLogSdesResponse(request.body)
-          (movementId, messageId) = extractMovementMessageId(sdesResponse)
-          persistenceResponse <- sdesResponse.notification match {
-            case SdesNotificationType.FileProcessed =>
-              updateStatus(movementId, messageId, MessageStatus.Success)
-            case SdesNotificationType.FileProcessingFailure =>
-              updateStatus(movementId, messageId, MessageStatus.Failed)
-            case _ => EitherT.rightT[Future, PresentationError]((): Unit)
-          }
-        } yield persistenceResponse).fold[Result](
-          error => Status(error.code.statusCode)(Json.toJson(error)),
-          _ => Ok
-        )
+        parseAndLogSdesResponse(request.body) match {
+          case Left(presentationError) =>
+            Future.successful(Status(presentationError.code.statusCode)(Json.toJson(presentationError)))
+          case Right(sdesResponse) =>
+            val (movementId, messageId) = extractMovementMessageId(sdesResponse)
+            (for {
+              persistenceResponse <- sdesResponse.notification match {
+                case SdesNotificationType.FileProcessed =>
+                  updateAndSendNotification(
+                    movementId,
+                    messageId,
+                    MessageStatus.Success,
+                    Json.toJson(
+                      Json.obj(
+                        "code" -> OK,
```

Comment:
What is `OK`? Fairly sure this will return 200, not "OK".

Is "OK" right here, or would something like "SUCCESS" or "PROCESSED" be better?

Reply:
Actually -- shouldn't this be the Hateoas response that we send from `common-transit-convention-traders`? What do we send on success if a trader sends a small message down the large message route and it succeeds? We need to do the same thing here.

Reply:
Here with this endpoint we don't know movementType then how we are going to create href with Hateoas? Also I have discussed this with you and you suggested to use like below so I have used that.
If SDES comes back with a FileProcessingSuccess message, we might need a new message for that, possibly something like:
{
    "code": "ACCEPTED"
    "message": "The message abcdef for movement 0123456 was successfully processed"
}

Reply:
> Also I had discussed this with you and you suggested to use like below so I have used that.

Sure, but now I've reviewed this I've realised we're going to be inconsistent by doing this, and I'd rather not be inconsistent.

> Here with this endpoint we don't know movementType then how we are going to create href with Hateoas?

We control the system, so if we need to figure out a way to do it, then we should.

> If SDES comes back with a FileProcessingSuccess message, we might need a new message for that

...that's what this is though, `FileProcessed` is the success message.

Reply:
I think we can get the movementType by passing movementType with SdesProperties while sending call to Sdes? or another solution I can see by calling transit-movements but not sure if we have endpoint for this on transit-movements without eori number. Please suggest.

Reply:
Let's take a step back a sec, because I'm thinking about this more and whether what's been done in `common-transit-convention-traders` is correct.

When we send a small message to the large message route, what is the success message we get pushed to us? I see you approved https://github.com/hmrc/common-transit-convention-traders/pull/415, did you look at what the overall result was?

Can you run that, and this, capture the push notification, and paste them here?

Reply:
I should have checked what the overall result should have been too, of course. I'm concerned that what we've got in the API layer and what we have here are inconsistent, and because we've not been able to test them in regression tests, I'm worried that the wrapping that `transit-movements-push-notifications` adds will make the Hateoas stuff redundant.

Either way, we're not being consistent now and we need to be.

Reply:
for router sdes call back(router):

``` {“messageUri”:“/customs/transits/movements/departures/64403215820425ef/messages/6440321586faaf71",“messageBody”:“{\“code\“:\“SUCCESS\“,\“message\“:\“The message 6440321586faaf71 for movement 64403215820425ef was successfully processed\“}”}  ```

for small message via large message route (api):

``` 
{“messageUri”:“/customs/transits/movements/departures/6440332a08662c97/messages/6440332a12db2b6e”,“messageBody”:“{\“_links\“:{\“self\“:{\“href\“:\“/customs/transits/movements/departures/6440332a08662c97/messages/6440332a12db2b6e\“},\“departure\“:{\“href\“:\“/customs/transits/movements/departures/6440332a08662c97\“}}}“}. ```

@drnaylor-oc   Please see the above notification and suggest.

Reply:
They're both wrong, no-one has really thought about this and so we have a lot of sub-optimal notifications.

Let's think about this and what traders get in a journey for a success.

**SMALL MESSAGE via SMALL MESSAGE API**

Sync response: Hateoas with box ID

**SMALL MESSAGE via LARGE MESSAGE API**

Sync response: Hateoas with box ID + upload information
Push notification: Hateoas that mirrors small message, wrapped in movement/message ID information

**LARGE MESSAGE via LARGE MESSAGE API**

Sync response: Hateoas with box ID + upload information
Push notification: Stringified JSON saying "SUCCESS" wrapped in movement/message ID information

---

Three routes, three different notifications. 

I admit, what I said about returing Hateoas in the large message/large message route doesn't make sense when you take all this in effect, so I apologise about that. However, I don't apologise for catching the inconsistency here -- this is why we should have been testing push notifications earlier because we might have realised this earlier.

I think there are two things that need to happen, and one not need to happen:

* Let's keep what you propose for large messages (SUCCESS etc.) -- that's fine
* We also need to alter the small message/large api route to also return "SUCCESS"
* `transit-movements-push-notifications` need to handle Json payloads differently, so the ultimate payload looks something like:

```json
{
    “messageUri”: “/customs/transits/movements/departures/64403215820425ef/messages/6440321586faaf71",
    "notificationType": "SENT_MESSAGE",
    “response”: {
        "code": "SUCCESS",
        "message": "The message 6440321586faaf71 for movement 64403215820425ef was successfully processed"
    }
} 
```

and for messages going to the trader

```json
{
    “messageUri”: “/customs/transits/movements/departures/64403215820425ef/messages/6440321586faaf71",
    "notificationType": "RECEIVED_MESSAGE",
    "messageBody": "<xml></xml>"
} 
```

By doing that, we make it super easy for traders to differentiate between a notification about a message, and an actual message coming back.

I'll put this into a ticket. For now, just concentrate on this ticket.

Reply:
So I hope everything is fine with this ticket just need to alter transit-movements-push-notifications to handle the above cases. right?

Reply:
> just need to alter transit-movements-push-notifications to handle the above cases. right?

Yes, and a slight change in `common-transit-convention-traders`. I've made tickets for them and passed them to Andy.